datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -224,6 +226,51 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
     }
 }
 
+#[allow(rustdoc::private_intra_doc_links)]
+/// Canonicalize any BinaryExprs that are not in canonical form
+/// <literal> <op> <col> is rewritten to <col> <op> <literal>
+/// <col1> <op> <col2> is rewritten so that the name of col1 sorts higher than col2 (b > a would be canonicalized to a < b)
+struct Canonicalizer {}
+
+impl Canonicalizer {
+    fn new() -> Self {
+        Self {}
+    }
+}
+
+impl TreeNodeRewriter for Canonicalizer {
+    type N = Expr;
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        if let Expr::BinaryExpr(BinaryExpr { left, op, right }) = expr {
+            match (left.as_ref(), right.as_ref(), op.swap()) {
```

Comment:
You can do something like this to reduce the indent of this function if you want (not required, I just figured I would point it out given you say you are looking to improve your rust skills

```suggestion
        let Expr::BinaryExpr(BinaryExpr { left, op, right }) = expr else {
          return Ok(expr)
        };
        match (left.as_ref(), right.as_ref(), op.swap()) {
```